### PR TITLE
VeTools Milestone 1 Submission

### PR DESCRIPTION
**Pull request by hazelcaus**
# Milestone Submission

* **PR Link:** https://github.com/vechain/grant-program/pull/151

All deliverables completed per agreement. Additionally, VeTools v1.0.0 is [live on the VSCode Marketplace](https://marketplace.visualstudio.com/items?itemName=magnusekstrom.vetools). Source code: https://github.com/hazelcaus/vetools

### CODE REVIEW

**Comment by hazelcaus** [comment]
@NecoSherry @libotony

**Comment by libotony** [comment]
Hi @hazelcaus, thank you for the milestone delivery, can you provide some info regarding the result of the delivery, e.g. screenshots, documents, etc?  And the source code seems to be a private repo, I can not access it.

File: milestone-delivery/deliveries/Vetools_milestone_1 [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# Vetools - Milestone Delivery :mailbox:
```
Comment (libotony) [inline milestone-delivery/deliveries/Vetools_milestone_1:1]: The filename should be `xxx.md`.

**Review by libotony — CHANGES_REQUESTED** [review]

**Comment by hazelcaus** [comment]
@libotony I have made the repo public!

**Comment by libotony** [comment]
> @libotony I have made the repo public!

Yes, I can see the code, I understand it's still the early stage of your project and there is not much document. To verify the milestone, could you provide some screenshots,videos, gifs for proving the delivery?

**Comment by hazelcaus** [comment]
Thanks @libotony for the suggestions. I have added some testing instructions and demos to the repo. Hope this helps!

**Comment by hazelcaus** [comment]
@libotony

**Comment by libotony** [comment]
> @libotony

Hi @hazelcaus thank you for the feedback, I have tried the first step, but I got an error 
Screenshot 2023-11-08 at 18 52 11

**Comment by hazelcaus** [comment]
That's strange @libotony. It's working fine for me. See: https://streamable.com/e/4yx8jz. Can you confirm if you're on the latest version?

**Comment by libotony** [comment]
> That's strange @libotony. It's working fine for me. See: https://streamable.com/e/4yx8jz. Can you confirm if you're on the latest version?

Yes, it's the latest version and I think it's the only version that published to the marketplace.

**Comment by hazelcaus** [comment]
Huh, I literally just recorded this video and it was working fine for me.

**Comment by libotony** [comment]
Extension Host is reporting `You don't have npm installed. Please install before continuing`

**Comment by hazelcaus** [comment]
Here's [another video](https://streamable.com/c7gzp7) I recorded demonstrating that it works properly. Also, could you take a screenshot(s) of the `Output -> VeTools` section (shown in the video section). Could help with some debugging... 
@libotony

**Comment by libotony** [comment]
> Here's [another video](https://streamable.com/c7gzp7) I recorded demonstrating that it works properly. Also, could you take a screenshot(s) of the `Output -> VeTools` section (shown in the video section). Could help with some debugging... @libotony

Only one line `vetools>>> 9.8.0`

**Comment by hazelcaus** [comment]
@libotony thanks for the feedback. I have made a change in the code where if the installation fails, it prompts you to manually run `npm install` or `yarn install`. Please update your extension version (to v1.0.1) and let me know if it works!

**Comment by libotony** [comment]
> @libotony thanks for the feedback. I have made a change in the code where if the installation fails, it prompts you to manually run `npm install` or `yarn install`. Please update your extension version (to v1.0.1) and let me know if it works!

Thanks for the update, I have successfully created a new vetools project. But I can not find the command `VeTools: Compile Contracts`.
Screenshot 2023-11-14 at 12 30 37

**Comment by hazelcaus** [comment]
My bad. I disabled that feature when testing for your bug, and kept it disabled in v1.0.1. Please update to `v1.0.2`, and you should be able to compile your contracts successfully. Thanks

**Comment by libotony** [comment]
> My bad. I disabled that feature when testing for your bug, and kept it disabled in v1.0.1. Please update to `v1.0.2`, and you should be able to compile your contracts successfully. Thanks

Thanks for the update, the compile command has shown after upgrading to `v1.0.2`. However, the command failed to execute.
Screenshot 2023-11-15 at 10 11 35

And output shows:
```
vetools>>> /bin/sh: node: command not found
```

**Comment by hazelcaus** [comment]
okay, I'm very confused as to why this error is visible for you. I've tried this on different machines (and OSes), but it works perfectly fine for me. What's happening under the hood of the extension is we spin up a child process and execute shell commands (like getting npm or node versions). With your initial error and the error you've got today, my guess is that `node` and `npm` are not added in your PATH correctly (particularly in a bash shell: `/bin/sh`). 

That said, could I request if you could install [Node Version Manager (nvm)](https://www.freecodecamp.org/news/node-version-manager-nvm-install-guide/) -- it could possibly help correct any faulty installations, PATH additions, and old versions. I've also released `v1.0.3` of the extension and added more debugging information. 

I apologize for the back-and-forth, but this is the fastest way to debug this issue(s) across different computers. Alternatively, I'm happy to send you a screen recording of the extension working across different systems.

**Comment by libotony** [comment]
> okay, I'm very confused as to why this error is visible for you. I've tried this on different machines (and OSes), but it works perfectly fine for me. What's happening under the hood of the extension is we spin up a child process and execute shell commands (like getting npm or node versions). With your initial error and the error you've got today, my guess is that `node` and `npm` are not added in your PATH correctly (particularly in a bash shell: `/bin/sh`).
> 
> That said, could I request if you could install [Node Version Manager (nvm)](https://www.freecodecamp.org/news/node-version-manager-nvm-install-guide/) -- it could possibly help correct any faulty installations, PATH additions, and old versions. I've also released `v1.0.3` of the extension and added more debugging information.
> 
> I apologize for the back-and-forth, but this is the fastest way to debug this issue(s) across different computers. Alternatively, I'm happy to send you a screen recording of the extension working across different systems.
> 
> Also, please kindly run the following in your terminal and confirm the versions for me (verify if you get any errors across different terminal sessions like ZSH, and BASH). Thanks!
> 
> ```shell
> npm --version
> node --version
> nvm version # if you've installed above
> ```

I have two MacBook for testing,

The first one, `nvm` installed, v1.0.3 tells `command 'vetools.newProject' not found` and `command 'vetools.buildContracts' not found`, `sh`/`bash`/`zsh` can show node version.

The second one, node is installed via brew, same as above, v1.0.3 tells `command 'vetools.newProject' not found` and `command 'vetools.buildContracts' not found`, `sh`/`bash`/`zsh` can show node version.

They both showing the same output:
```
try_execute >>> workingDirectory: undefined
commands: node
args: --version

vetools>>> /bin/sh: node: command not found
check_apps>>> Current state for versions: [{"app":"node","isValid":false,"required_version":{"max":"17.0.0","min":"14.0.0"},"version":""}] Invalid: true
```

**Comment by hazelcaus** [comment]
Just to be clear: in v1.0.3, "New Project" and "Compile Contracts" both don't work? Or just the latter?

**Comment by libotony** [comment]
> Just to be clear: in v1.0.3, "New Project" and "Compile Contracts" both don't work? Or just the latter?
> 
> Here's my output (also on a Mac) Screenshot 2023-11-15 at 10 06 22

Both don't work, I also asked my colleagues to verify.

**Comment by hazelcaus** [comment]
@libotony I might be on to something (and figured out the error). Pushing a fix now...

**Comment by hazelcaus** [comment]
@libotony please update to v1.0.4 and let me know if you're able to compile contracts successfully. Also, would suggest you start over with `VeTools: New Project` if you're not already doing that. Made some changes in v1.0.2 that might cause an error when compiling contracts

**Comment by libotony** [comment]
> @libotony please update to v1.0.4 and let me know if you're able to compile contracts successfully. Also, would suggest you start over with `VeTools: New Project` if you're not already doing that. Made some changes in v1.0.2 that might cause an error when compiling contracts

Thank you for the update, it's working perfectly!

**Review by libotony — APPROVED** [review]

**Comment by libotony** [comment]
I noticed that you are using `@vechain.energy/hardhat-thor` in the project, it's deprecated due to the foundation published a new one, please take look https://github.com/vechainfoundation/hardhat-plugins

**Comment by hazelcaus** [comment]
noted. This has been reflected in `v1.0.5`

**Review by NecoSherry — APPROVED** [review]

**Pull request merged**